src/sagas/vmChanges.js

```diff
@@ -177,54 +162,132 @@ function* composeProvisionSourceTemplate ({ vm, basic, disks }) {
-      }
+    return Object.keys(changes).length === 0
+      ? false
+      : { id: disk.id, ...changes }
+  })
 
-      // did the storage domain change?
-      if (disk.storageDomainId !== templateDisk.get('storageDomainId')) {
-        diskRequiresClone = true
-        changesToTemplateDisk.storage_domains = {
-          storage_domain: [{ id: disk.storageDomainId }],
-        }
-      }
+  /*
+   * If we change ANY disk on the template, we need to send in ALL disks.
+   */
+  if (diskChanges.some(Boolean)) {
+    vmUpdates.disk_attachments = {
+      disk_attachment: diskChanges.map(
+        (changed, index) => ({
+          disk: changed || { id: disksFromTemplate[index].disk.id },
+        })
+      ),
+    }
+  }
 
-      if (Object.keys(changesToTemplateDisk).length > 1) {
-        if (vmUpdates.disk_attachments) {
-          // add another disk to clone
-          vmUpdates.disk_attachments.disk_attachment.push({ disk: changesToTemplateDisk })
-        } else {
-          // initial setup a disk to clone
-          vmUpdates.disk_attachments = {
-            disk_attachment: [{
-              disk: changesToTemplateDisk,
-            }],
-          }
-        }
-      }
-    })
+  return [vmUpdates, storageAllocation === 'clone']
+}
+
+/**
+ * Select the disk's __format__ and __sparse__ based on where the disk is coming from,
+ * what changes the user has made to the VM from the template, and what the disk's
+ * target storage domain looks like.
+ *
+ * Webadmin behavior:
+ *   [Template OptimizedFor -> VM OptimizedFor -> SD changes? ~~ Storage allocation ... disk format/sparse outcome]
+ *   (* = default)
+ *
+ *   Server  -> Server*  -> No SD changes* ~~ Clone ... format stays the same, sparse stays the same
+ *   Server  -> Server*  -> SD changes ~~ Clone     ... format stays the same, sparse changes as necessary based on format + SD
+ *   Server  -> Desktop  -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Server  -> Desktop  -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Desktop* -> No SD changes* ~~ Thin  ... format: cow, sparse: true
+ *   Desktop -> Desktop* -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> No SD changes* ~~ Clone ... format: cow, sparse: true (Thin settings retained)
+ *   Desktop -> Server   -> SD changes ~~ Clone     ... format: cow, sparse: true (Thin settings retained)
+ */
+function determineTemplateDiskFormatAndSparse (templateOptimizedFor, vmOptimizedFor, baseDisk, userDisk, storageAllocation, targetSD) {
+  const attributes = {}
+
+  if (storageAllocation === 'thin') {
+    attributes.format = 'cow'
+    attributes.sparse = targetSD.getIn(['defaultDiskFormatToSparse', 'cow'])
+  }
+
+  if (templateOptimizedFor === 'desktop' || vmOptimizedFor === 'desktop') {
+    attributes.format = 'cow'
+    attributes.sparse = targetSD.getIn(['defaultDiskFormatToSparse', 'cow'])
```

Comment:
not sure if we should force `cow` format if the domain does not support it i.e. for for file domains `targetSD.diskAttributesForDiskType.thin.format ==  'raw'`  and  `targetSD.defaultDiskFormatToSparse.cow = true` (as for every domain) and `sparse == thin`.

Reply:
It is more complicated than that in webadmin actually.  File domains are prefectly happy with raw/sparse, raw/!sparse, cow/sparse, cow/!sparse all based on more specific options than what is allowed via web-ui.  For example, the incremental backup checkbox, when checked, will force the disk to be cow.

I'll take another look as there are lots of questions in this code set anyway.

Reply:
OK. So those are different (not symmetric) mappings.
Just to confirm: 
1. `targetSD.diskAttributesForDiskType.thin.format == 'raw'`  means that `raw` format is the default (not the only supported) format for `thin` allocation policy for  "Add new disk" scenario? 
2. `targetSD.defaultDiskFormatToSparse.cow = true` means that for template disks with `cow` format we should use `thin` allocation policy?

Reply:
> OK. So those are different (not symmetric) mappings.
> Just to confirm:
> 
>   1. `targetSD.diskAttributesForDiskType.thin.format == 'raw'`  means that `raw` format is the default (not the only supported) format for `thin` allocation policy for  "Add new disk" scenario?

Correct.

> 
>     2. `targetSD.defaultDiskFormatToSparse.cow = true` means that for template disks with `cow` format we should use `thin` allocation policy?

Essentially true.  It means template disks with `cow` format in the target storage domain should be `sparse: true`.

The diskType / allocation policy is a UI construct that maps to both format and sparse.  While highly correlated,  I don't consider them interchangeable terms due to the weird rules that come up with template disks on the resource allocation tab in webadmin create vm.